site-manage/site-manage-impl/impl/src/java/org/sakaiproject/sitemanage/impl/SiteManageServiceImpl.java

```diff
@@ -698,6 +704,54 @@ protected void doInTransactionWithoutResult(TransactionStatus status) {
         return transversalMap;
     }
 
+    private Set<String> getToolPermissionCandidatesToCopy(String siteId, Map<String, List<String>> importTools, Map<String, Map<String, List<String>>> toolOptions) {
+
+		Set<String> toolIds = (importTools != null) ? new HashSet<>(importTools.keySet()) : Collections.emptySet();
+		if (toolOptions != null) {
+			toolIds.addAll(toolOptions.keySet()); // the instructor might want to copy permissions without any other content from the tool
+		}
+
+		Set<String> toolPermissions = new HashSet<>();
+		for (String toolId : toolIds) {
+
+			// set copyPermissionsSelected to true if a copy permissions option is selected
+			boolean copyPermissionsSelected = false;
+			if (toolOptions.containsKey(toolId)) {
+				Map<String, List<String>> options = toolOptions.get(toolId);
+				if (options.containsKey(siteId)) {
+					List<String> siteOptions = options.get(siteId);
+					copyPermissionsSelected = siteOptions.contains(EntityTransferrer.COPY_PERMISSIONS_OPTION);
+				}
+			}
+
+			for (EntityProducer ep : entityManager.getEntityProducers()) {
+				if (ep instanceof EntityTransferrer) {
+					try {
+						EntityTransferrer et = (EntityTransferrer) ep;
+						// if this producer claims this tool id
+						if (ArrayUtil.contains(et.myToolIds(), toolId)) {
+							// if the copy permission option was explicitly selected
+							// or the tool does not have a relevant Permissions panel configurable by an instructor
+							if (copyPermissionsSelected || !et.supportsTransferOption(EntityTransferrer.COPY_PERMISSIONS_OPTION)) {
+								String prefix = et.getToolPermissionsPrefix();
+								if (prefix != null) { // Only fetch registered functions that are tool specific
+									List<String> perms = functionManager.getRegisteredFunctions(et.getToolPermissionsPrefix());
+									if (perms != null && ! perms.isEmpty()) {
+										toolPermissions.addAll(perms);
+									}
+								}
+							}
+							break;
+						}
+					} catch (Exception e) {
+						log.error("Error encountered: [{}]", e);
```

Comment:
likely will want the stack trace if this fails

```suggestion
						log.error("could not copy permissions encountered", e);
```